sale_order_type/models/sale.py

```diff
@@ -23,20 +22,23 @@ def onchange_partner_id(self):
         if sale_type:
             self.type_id = sale_type
 
-    @api.multi
     @api.onchange("type_id")
     def onchange_type_id(self):
+        vals = {}
         for order in self:
-            if order.type_id.warehouse_id:
-                order.warehouse_id = order.type_id.warehouse_id
-            if order.type_id.picking_policy:
-                order.picking_policy = order.type_id.picking_policy
-            if order.type_id.payment_term_id:
-                order.payment_term_id = order.type_id.payment_term_id.id
-            if order.type_id.pricelist_id:
-                order.pricelist_id = order.type_id.pricelist_id.id
-            if order.type_id.incoterm_id:
-                order.incoterm = order.type_id.incoterm_id.id
+            vals.update(
+                {
+                    "warehouse_id": order.type_id.warehouse_id.id
+                    or order.warehouse_id.id,
+                    "picking_policy": order.type_id.picking_policy
+                    or order.picking_policy,
+                    "payment_term_id": order.type_id.payment_term_id.id
+                    or order.payment_term_id,
+                    "pricelist_id": order.type_id.pricelist_id.id or order.pricelist_id,
+                    "incoterm": order.type_id.incoterm_id.id or order.incoterm,
+                }
+            )
+            order.write(vals)
```

Comment:
The good method is not doing a write() in an onchange.

Either you affect each value (like in the code you deleted), but it has disadvantage to do a write after each affectation (and an SQL query).

Either (the best) you use update() on the record:

```python
order.update({
                    "warehouse_id": order.type_id.warehouse_id.id
                    or order.warehouse_id.id,
                    "picking_policy": order.type_id.picking_policy
                    or order.picking_policy,
                    "payment_term_id": order.type_id.payment_term_id.id
                    or order.payment_term_id,
                    "pricelist_id": order.type_id.pricelist_id.id or order.pricelist_id,
                    "incoterm": order.type_id.incoterm_id.id or order.incoterm,
                })
```

Moreover the way you did the update is a llitle bit tricky. Don't reaffect the order field value if the type value does not contain one, but build your dict with value or not.

So, I would have done:

```python
for order in self:
    vals = {}
    order_type = order.type_id
    if order_type.warehouse_id:
        vals.update({"warehouse_id": o_type.warehouse_id})
    if order_type.picking_policy:
        vals.update({"picking_policy": o_type.picking_policy})
    if order_type.payment_term_id:
        vals.update({"payment_term_id": o_type.payment_term_id})
    if order_type.pricelist_id:
        vals.update({"pricelist_id": o_type.pricelist_id})
    if order_type.incoterm_id:
        vals.update({"incoterm": order_type.incoterm_id})
    
    if vals:
        order.update(vals)
```

Reply:
@rousseldenis I fixed as your advice, thank you for review.